### PR TITLE
GH-36483: [C++] Make `UTF8StringToUTF16` and `UTF16StringToUTF8` accept `string_views`

### LINKED ISSUE #36483: [C++] Make `UTF8StringToUTF16` and `UTF16StringToUTF8` accept `string_views`
### Describe the enhancement requested

`UTF8StringToUTF16` and `UTF16StringToUTF8` accept `const` references as inputs. As @kou pointed out, these functions should accept `string_views`. 

### Component(s)

C++

### PR DESCRIPTION
**Pull request by sgilmore10**
### Rationale for this change

As @kou pointed out in #36366, it's better to pass read-only string data via  as string_views than as const references. I recently submitted two new functions called `UTF8StringToUTF16` and `UTF16StringToUTF8`, both of which accept const references. This PR changes the signatures of those functions to accept string_views.

### What changes are included in this PR?

Changed the function signatures of `UTF8StringToUTF16` and `UTF16StringToUTF8`:

1. `arrow::Result<std::u16string> UTF8StringToUTF16(const std::string& source)` ->
    `arrow::Result<std::u16string> UTF8StringToUTF16(std::string_view source)` 

2. `arrow::Result<std::string> UTF16StringToUTF8(const std::u16string& source)` ->
    `arrow::Result<std::string> UTF16StringToUTF8(std::u16string_view source)` 

### Are these changes tested?

Uses existing tests in `utf8_util_test.cc`

### Are there any user-facing changes?

No.

Only the MATLAB Interface uses these functions, so  I don't expect these changes to break anyone's code:

1. Search [results](https://github.com/search?q=repo%3Aapache%2Farrow%20UTF8StringToUTF16&type=code) for `UTF8StringToUTF16`
2. Search [results](https://github.com/search?q=repo%3Aapache%2Farrow+UTF16StringToUTF8&type=code) for `UTF16StringToUTF8`

* Closes: #36483

### CODE REVIEW

**Comment by sgilmore10** [comment]
I believe the failures are unrelated to the changes in this PR.

File: matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc [hunk 1]
```diff
@@ -49,7 +49,8 @@ namespace arrow::matlab::array::proxy {
         const mda::TypedArray<mda::MATLABString> units_mda = opts[0]["TimeUnit"];
 
         // extract the time zone string
-        MATLAB_ASSIGN_OR_ERROR(const auto timezone, arrow::util::UTF16StringToUTF8(timezone_mda[0]),
+        const std::u16string u16_timezone = timezone_mda[0];
```
Comment (kou) [inline matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc:52]: I don't know the type of `timezone_mda[0]` but does this cause a copy?

Can we use `std::u16string_view` here?

```cpp
std::u16string_view u16_timezone(timezone_mda[0]);
```

Comment (sgilmore10) [reply matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc:52]: Hi @kou, this does cause a copy. Unfortunately, it's not  possible to create a `std::816string_view` from `timezone_mda` directly. However, I can extract a `const` reference from `timezone_mda:

```cpp
const std::u16string& u16_timezone = timezone_mda[0];
```

That should avoid the copy.

Comment (sgilmore10) [reply matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc]: Just pushed the change.

**Review by kou — APPROVED** [review]
+1

**Pull request merged**